File: 0559-maximum-depth-of-n-ary-tree/0559-maximum-depth-of-n-ary-tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    int maxD = 1;
    void findMaxDepth(Node* node,int currDepth) {
        //recursively travel each child node and maximize maxD
        for(Node* child : node->children) {
            findMaxDepth(child,currDepth+1);
        }
        //maximize Depth
        if(maxD < currDepth) {
            maxD = currDepth;
        }
    }
    
    int maxDepth(Node* root) {
        //Base Case
        if(!root) return 0;
        findMaxDepth(root,1);
        return maxD;
    }
};

File: 0559-maximum-depth-of-n-ary-tree/0559-maximum-depth-of-n-ary-tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    int maxH = 1;
    public void findMaxH(Node node,int ht) {
        for(Node child:node.children) {
            findMaxH(child,ht+1);
        }
         if(node.children.size() == 0) {
            if(ht>maxH) {
                maxH = ht;
            }
         }
    }
    public int maxDepth(Node root) {
        if(root==null) return 0;
        findMaxH(root,maxH);
        return maxH;
    }
}